src/orchestrator/chat_completions_detection.rs

```diff
@@ -1,27 +1,487 @@
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+use axum::http::HeaderMap;
+use futures::{
+    future::{join_all, try_join_all},
+    Future,
+};
+use std::{collections::HashMap, sync::Arc};
+use std::{
+    pin::Pin,
+    time::{SystemTime, UNIX_EPOCH},
+};
+use tracing::{debug, info, instrument};
+
+use super::{ChatCompletionsDetectionTask, Context, Error, Orchestrator};
+use crate::{
+    clients::{
+        detector::{ChatDetectionRequest, ContentAnalysisRequest},
+        openai::{
+            ChatCompletion, ChatCompletionChoice, ChatCompletionsRequest, ChatCompletionsResponse,
+            Content, DetectionResult, InputDetectionResult, OpenAiClient,
+        },
+    },
+    config::DetectorType,
+    models::{DetectorParams, OrchestratorDetectionResult},
+    orchestrator::{
+        detector_processing::content,
+        unary::{chunk, detect_content},
+        Chunk,
+    },
+};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+pub type ChunkResult<T> = Pin<Box<dyn Future<Output = T> + Send>>;
+
+/// Internal structure to capture chat messages (both request and response)
+/// and prepare it for processing
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct ChatMessageInternal {
+    // Index of the message
+    pub message_index: usize,
+    /// The role of the messages author.
+    pub role: String,
+    /// The contents of the message.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub content: Option<Content>,
+    /// The refusal message by the assistant. (assistant message only)
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub refusal: Option<String>,
+}
+
+pub type ChatMessagesInternal = Vec<ChatMessageInternal>;
+pub enum DetectorRequest {
+    ContentAnalysisRequest(ContentAnalysisRequest),
+    ChatDetectionRequest(ChatDetectionRequest),
+}
+
+// Get ChatMessagesInternal from ChatCompletionsRequest
+impl From<ChatCompletionsRequest> for ChatMessagesInternal {
+    fn from(value: ChatCompletionsRequest) -> Self {
+        let mut messages = ChatMessagesInternal::new();
+        value.messages.iter().enumerate().for_each(|(index, m)| {
+            messages.push({
+                ChatMessageInternal {
+                    message_index: index,
+                    role: m.role.clone(),
+                    content: m.content.clone(),
+                    refusal: m.refusal.clone(),
+                }
+            })
+        });
+        messages
+    }
+}
```

Comment:
This is inefficient; allocates every `for_each` iteration as `Vec::new()` was used instead of `Vec::with_capacity(n)` and it creates unnecessary copies of the message contents. Use `into_iter()` and `map()`:

```suggestion
impl From<ChatCompletionsRequest> for ChatMessagesInternal {
    fn from(value: ChatCompletionsRequest) -> Self {
        value
            .messages
            .into_iter()
            .enumerate()
            .map(|(index, message)| ChatMessageInternal {
                message_index: index,
                role: message.role,
                content: message.content,
                refusal: message.refusal,
            })
            .collect()
    }
}
```